cis194/week7/daniel-deng/src/CIS194/Homework07/Exercise04.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module CIS194.Homework07.Exercise04 where
+
+import CIS194.Homework07.Buffer     ( Buffer(..) )
+import CIS194.Homework07.Scrabble   ( Score(..), getScore )
+import CIS194.Homework07.Sized      ( Size(..), getSize, size )
+import CIS194.Homework07.JoinList   ( JoinList(..), tag )
+import CIS194.Homework07.Exercise02 ( indexJ, dropJ, takeJ )
+import CIS194.Homework07.Exercise03 ( scoreString )
+
+type JoinListBuffer = JoinList (Score, Size) String
+
+jlToList :: Monoid m => JoinList m a -> [a]
+jlToList Empty              = []
+jlToList (Single _ a)       = [a]
+jlToList (Append _ jl1 jl2) = jlToList jl1 ++ jlToList jl2
+
+instance Buffer JoinListBuffer where
+
+  -- | Convert a buffer to a String.
+  toString = unlines . jlToList
+
+  -- | Create a buffer from a String.
+  fromString = mconcat . fmap toSingle . lines
```

Comment:
@shineli1984 I don't know why `mconcat` only supports a`[a]` as input, instead of something more generic, like `Foldable`? Is `mconcat :: [a] -> a` that a Category Theory thing or is it more a Haskell-specific API design?